objc/notify.m

```diff
@@ -1,74 +1,153 @@
-        NSAppleScript *findScript = [[NSAppleScript alloc] initWithSource:findString];
-        NSAppleEventDescriptor *resultDescriptor = [findScript executeAndReturnError:nil];
-        return [resultDescriptor stringValue];
+NSString* getBundleIdentifier(NSString* appName)
+{
+    NSString* findString = [NSString stringWithFormat:@"get id of application \"%@\"", appName];
+    NSAppleScript* findScript = [[NSAppleScript alloc] initWithSource:findString];
+    NSAppleEventDescriptor* resultDescriptor = [findScript executeAndReturnError:nil];
+    return [resultDescriptor stringValue];
 }
 
 // setApplication(new_bundle_identifier: &str) -> Result<()>
-BOOL setApplication(NSString *newbundleIdentifier) {
-        if(LSCopyApplicationURLsForBundleIdentifier((CFStringRef)newbundleIdentifier, NULL) != NULL) {
-                fakeBundleIdentifier = newbundleIdentifier;
-                return YES;
-        }
-        return NO;
+BOOL setApplication(NSString* newbundleIdentifier)
+{
+    if (LSCopyApplicationURLsForBundleIdentifier((CFStringRef)newbundleIdentifier, NULL) != NULL)
+    {
+        fakeBundleIdentifier = newbundleIdentifier;
+        return YES;
+    }
+    return NO;
 }
 
-// scheduleNotification(title: &str, subtitle: &str message: &str, sound: &str, f64) -> NotificationResult<()>
-bool scheduleNotification(NSString *title, NSString *subtitle, NSString *message, NSString *sound, double deliveryDate) {
-        @autoreleasepool {
-                if (!installNSBundleHook()) {
-                        return NO;
-                }
-                NSDate *scheduleTime = [NSDate dateWithTimeIntervalSince1970:deliveryDate];
-                NSUserNotificationCenter *nc = [NSUserNotificationCenter defaultUserNotificationCenter];
-                NotificationCenterDelegate *ncDelegate = [[NotificationCenterDelegate alloc] init];
-                ncDelegate.keepRunning = YES;
-                nc.delegate = ncDelegate;
-
-                NSUserNotification *note = [[NSUserNotification alloc] init];
-                note.title = title;
-                if (![subtitle isEqualToString:@""]) {
-                        note.subtitle = subtitle;
-                }
-                note.informativeText = message;
-                note.deliveryDate = scheduleTime;
-                if (![sound isEqualToString:@"_mute"]) {
-                        note.soundName = sound;
-                }
-                [nc scheduleNotification:note];
-                [NSThread sleepForTimeInterval:0.1f];
-                return YES;
+// sendNotification(title: &str, subtitle: &str, message: &str, options: NotificationOptions) -> NotificationResult<()>
+NSDictionary* sendNotification(NSString* title, NSString* subtitle, NSString* message, NSDictionary* options)
+{
+    @autoreleasepool
+    {
+        if (!installNSBundleHook())
+        {
+            // TODO: Could potentially have different error messages
+            return @{@"error" : @""};
```

Comment:
@hoodie Just fyi, that's a `NSDictionary` literal, not just an empty string. Your question still applies though.

Reply:
Yes I'm using an NSDictionary. The reason why I'm passing an empty string and not a boolean is explained in my initial comment of the PR.
> I'm unsure if this is a limitation of the objective c to rust library, or if I missed something, but I was unable to create an NSDictionary with multiple value types. As a result, I was forced to stringify arrays, numbers, and booleans to pass them to the objective c side but would prefer not to.

Basically, when going to and from rust, I'm using `NSDictionary<NSString, NSString>` because I don't know any way to accept any value type (pseudocode: `NSDictionary<NSString, NSString | bool | NSArray>`).

I was thinking I could perhaps use some kind of union (the C kind, not rust enums), but I'm not sure if/how that would work.

Reply:
there are c like `unions` in rust, but I think they're there explicitly for c-libraries that expect them. Also seems like the objc-foundation crate is a bit stale. I have no clue if this can be optimized, but this seems to be a reasonable solution. we're sending one at a time notifications for humans here, there is no particular need for efficiency or high throughput, so picking a high level abstraction is very acceptable to :+1: